File: asml/LC554BrickWall.py
class Solution(object):
    def leastBricks(self, wall):
        if not wall:
            return 0
        dictionary = {}
        for each in wall:
            length = 0
            for brick in each[:-1]:
                dictionary[length+brick] = dictionary.get(length+brick, 0)+1
                length += brick
        return len(wall) - max(dictionary.values()+[0])

class Solution(object):
    def leastBricks(self, walls):
        if not walls:
            return 0
        dictionary = {}
        for wall in walls:
            length = 0
            for brick in wall[:-1]:
                dictionary[length+brick] = dictionary.get(length+brick, 0)+1
                length += brick
        if not dictionary:
            return len(walls)
        else:
            return len(walls) - max(dictionary.values())

        

File: asml/LC554BrickWall.cpp
class Solution {
public:
    int leastBricks(vector<vector<int>>& wall) {
        if (wall.empty()) return 0;
        unordered_map<int, int> dictionary;
        int row = wall.size();
        int minimum = 0;
        for (int i = 0; i < row; i++) {
            int length = 0;
            for (int j = 0; j < wall[i].size()-1; j++) {
                length += wall[i][j];
                dictionary[length]--;
                minimum = min(minimum, dictionary[length]);
            }
        }
        return row+minimum;
    }
};
